sdk/typespec/typespec_derive/tests/compilation-tests.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+use std::{collections::HashMap, io::BufReader, path::PathBuf, process::Stdio};
+
+use cargo_metadata::{diagnostic::DiagnosticLevel, Message};
+use serde::Serialize;
+
+#[derive(Serialize)]
+struct MessageExpectation {
+    pub level: DiagnosticLevel,
+    pub code: Option<String>,
+    pub message: Option<String>,
+    pub spans: Vec<MessageSpan>,
+}
+
+#[derive(Serialize)]
+struct MessageSpan {
+    pub file_name: String,
+    pub line: usize, // We only check the line number of the span because other properties (like highlight span) can vary by compiler version.
+}
+
+#[derive(Serialize)]
+#[serde(transparent)]
+struct FileResult {
+    pub messages: Vec<MessageExpectation>,
+}
+
+#[test]
+pub fn compilation_tests() {
+    let test_root = {
+        let mut cwd = std::env::current_dir().expect("failed to get current directory");
```

Comment:
Is this documented - stable - that the PWD during `cargo test` is the parent directory? Otherwise, you might consider using the `file!()` macro here instead.

As for the "compilation-tests" directory, to avoid confusion that it's an actual crate, could we put it in "tests/data"? We have a fairly common pattern across many languages of this and could add blanket exemptions for directories of that pattern for various things.

Reply:
I changed this to use `CARGO_MANIFEST_DIR`, which is more stable and gives us what we want.